pkg/rp/util/authClient/azureWorkloadIdentity.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2023 The Radius Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authClient
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/containers/azcontainerregistry"
+	"github.com/radius-project/radius/pkg/to"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/retry"
+)
+
+var _ AuthClient = (*azureWorkloadIdentity)(nil)
+
+type azureWorkloadIdentity struct {
+	clientID string
+	tenantID string
+}
+
+func NewAzureWorkloadIdentity(clientID string, tenantID string) AuthClient {
+	return &azureWorkloadIdentity{clientID: clientID, tenantID: tenantID}
+}
+
+func (b *azureWorkloadIdentity) GetAuthClient(ctx context.Context, templatePath string) (remote.Client, error) {
+	c := azcontainerregistry.AuthenticationClientOptions{
+		azcore.ClientOptions{
+			Retry: policy.RetryOptions{
+				MaxRetries: 10,
+			},
+		},
+	}
+
+	opt := &azidentity.WorkloadIdentityCredentialOptions{
+		ClientID: b.clientID,
+		TenantID: b.tenantID,
+	}
+
+	cred, err := azidentity.NewWorkloadIdentityCredential(opt)
+	if err != nil {
+		return nil, err
+	}
+
+	aadToken, err := cred.GetToken(ctx, policy.TokenRequestOptions{
+		Scopes: []string{"https://containerregistry.azure.net/.default"}})
+	if err != nil {
+		return nil, err
+	}
+
+	registryHost, err := getRegistryHostname(templatePath)
+	if err != nil {
+		return nil, err
+	}
+	ac, err := azcontainerregistry.NewAuthenticationClient(fmt.Sprintf("https://%s", registryHost), &c)
+	if err != nil {
+		return nil, err
+	}
+
+	rt, err := ac.ExchangeAADAccessTokenForACRRefreshToken(ctx, "access_token", registryHost, &azcontainerregistry.AuthenticationClientExchangeAADAccessTokenForACRRefreshTokenOptions{
```

Comment:
did you reference this from a docs page somewhere? if so, maybe we should link it here so that future work here will take less ramp-up time

Reply:
azure container registry team have not released it yet, its part of their main repo. I will add the referent once the next release of [azcontainerregistry](https://github.com/Azure/azure-sdk-for-go/blob/c3ab67ad51c8a13622116ce057d1ce3cd2656f9e/sdk/containers/azcontainerregistry/authentication_client.go#L37) is out